components/script/dom/window.rs

```diff
@@ -2072,6 +2076,59 @@ impl Window {
         issued_reflow
     }
 
+    /// If parsing has taken a long time and reflows are still waiting for the `load` event,
+    /// start allowing them. See <https://github.com/servo/servo/pull/6028>.
+    pub(crate) fn reflow_if_reflow_timer_expired(&self, can_gc: CanGc) {
+        // Only trigger a long parsing time reflow if we are in the first parse of `<body>`
+        // and it started more than `INITIAL_REFLOW_DELAY` ago.
+        if !matches!(
+            self.layout_blocker.get(),
+            LayoutBlocker::Parsing(instant) if instant + INITIAL_REFLOW_DELAY < Instant::now()
+        ) {
+            return;
+        }
+        self.allow_layout_if_necessary(can_gc);
+    }
+
+    /// Block layout for this `Window` until parsing is done. If parsing takes a long time,
+    /// we want to layout anyway, so schedule a moment in the future for when layouts are
+    /// allowed even though parsing isn't finished and we havne't sent a load event.
+    pub(crate) fn prevent_layout_until_load_event(&self) {
+        // If we have already started parsing or have already fired a load event, then
+        // don't delay the first layout any longer.
+        if !matches!(self.layout_blocker.get(), LayoutBlocker::WaitingForParse) {
+            return;
+        }
+
+        self.layout_blocker
+            .set(LayoutBlocker::Parsing(Instant::now()));
+    }
+
+    /// Inform the [`Window`] that layout is allowed either because `load` has happened
+    /// or because parsing the `<body>` took so long that we cannot wait any longer.
+    pub(crate) fn allow_layout_if_necessary(&self, can_gc: CanGc) {
+        if matches!(
+            self.layout_blocker.get(),
+            LayoutBlocker::FiredLoadEventOrParsingTimerExpired
+        ) {
+            return;
+        }
+
+        self.layout_blocker
+            .set(LayoutBlocker::FiredLoadEventOrParsingTimerExpired);
+        self.Document().set_needs_paint(true);
+
+        // We do this immediately instead of scheduling a future task, because this can
+        // happen if parsing is taking a very long time, which means that the
+        // `ScriptThread` is busy doing the parsing and not doing layouts.
+        //
+        // TOOD(mrobinson): It's expected that this is necessary when in the process of
+        // parsing, as we need to interrupt it to update contents, but why is this
+        // necessary when parsing finishes? Not doing the synchronous update in that case
+        // causes iframe tests to become flaky.
+        with_script_thread(|script_thread| script_thread.update_the_rendering(false, can_gc))
```

Comment:
I'm not sure it is a good idea to run the full update here, as opposed to only doing a reflow, because [finishing parsing a script tag](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-incdata) seems to come with it's own approach to performing a microtask checkpoint. Would only doing a reflow here be sufficient? If so, I think it would be the approach with less possible side-effects, either now or if we make changes to the parser(which seems to lag behind the spec from a quick glance).

Reply:
How would you feel about running *update the rendering*, but without running a microtask checkpoint? I'm a bit worried about doing reflow of a single pipeline these days.  I think it might be leading to early exits of the test runner / flaky result and I'd like to handle this a bit more thoroughly in *update the rendering*.

Reply:
Scratch that. I think the issue is related to our problems with `<iframe>` sizes. I'm working on that now and should hopefully be able to remove this final reflow.